torch/csrc/jit/ir.cpp

```diff
@@ -250,6 +254,220 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    }
+  } else {
+    auto name = val->readableName();
+
+    auto aliased_name = aliases.find(unique);
+    if (aliased_name != aliases.end()) {
+      name = aliased_name->second;
+    }
+
+    if (isdigit(name.at(0))) {
+      out << "%";
+    }
+    out << name;
+  }
+return out;
+}
+
+std::ostream& prettyPrintInputs(
+  std::ostream & out,
+  const Node* node,
+  constant_list& constants,
+  alias_list& aliases
+) {
+  out << "(";
+  auto delimiter = "";
+  for (const auto* in_value : node->inputs()) {
+    out << delimiter;
+    prettyPrintValue(out, in_value, constants, aliases);
+    delimiter = ", ";
+  }
+  out << ")";
+  return out;
+}
+
+std::ostream& prettyPrintNodeVisitor(
+  std::ostream & out,
+  const Node* root,
+  int level,
+  constant_list& constants,
+  alias_list& aliases
+);
+
+std::ostream& prettyPrintBlockVisitor(
+  std::ostream & out,
+  const Block* root,
+  int level,
+  constant_list& constants,
+  alias_list& aliases
+) {
+  for (const auto* node : root->nodes()) {
+    prettyPrintNodeVisitor(out, node, level, constants, aliases);
+  }
+
+  prettyPrintNodeVisitor(out, root->return_node(), level, constants, aliases);
+
+  return out;
+}
+
+std::ostream& prettyPrintIf(
+  std::ostream & out,
+  const Node* node,
+  int level,
+  constant_list& constants,
+  alias_list& aliases
+) {
+  out << "if ";
+  auto if_block = node->blocks()[0];
+  auto else_block = node->blocks()[1];
+  prettyPrintValue(out, node->inputs()[0], constants, aliases);
+  out << ":" << std::endl;
+
+  // Print node contents
+  prettyPrintBlockVisitor(out, if_block, level + 1, constants, aliases);
+  // Print if block output
+  int if_count = 0;
+  for (const auto* output_value : node->outputs()) {
+    indent(out, level + 1);
+    prettyPrintValue(out, output_value, constants, aliases);
+    out << " = ";
+    prettyPrintValue(out, if_block->return_node()->inputs()[if_count++], constants, aliases);
+    out << std::endl;
+    if_count++;
+  }
+
+  indent(out, level);
+  out << "else:" << std::endl;
+  prettyPrintBlockVisitor(out, else_block, level + 1, constants, aliases);
+  int else_count = 0;
+  for (const auto* output_value : node->outputs()) {
+    indent(out, level + 1);
+    out << output_value->readableName();
+    out << " = ";
+    prettyPrintValue(out, else_block->return_node()->inputs()[else_count], constants, aliases);
+    out << std::endl;
+    else_count++;
+  }
+
+  return out;
+}
+
+std::ostream& prettyPrintLoop(
+  std::ostream & out,
+  const Node* node,
+  int level,
+  constant_list& constants,
+  alias_list& aliases
+) {
+  out << "while ";
+  prettyPrintValue(out, node->inputs()[1], constants, aliases);
+
+  out << std::endl;
+  auto body_block = node->blocks()[0];
+
+  // Offset to skip loop trip count
+  int count = 1;
+  for (const auto* block_param : body_block->param_node()->outputs())  {
+    // add names to map
+    aliases[block_param->uniqueName()] = node->inputs()[count]->readableName();
```

Comment:
So... what happens if the loop carried dependency is written before all uses of the loop input are used? I'm thinking about the non-pretty-printed case of:

```
While(cond, carry_init)
  graph (carry_in) {
    carry_out = Op(carry_in)
    new_cond = Op2(carry_in)
    return (new_cond, carry_out)
  }
```

Won't your aliasing logic turn this into:

```
carry_init = ...
while cond:
  carry_init = Op(carry_init)
  cond = Op2(carry_init)
```

That looks bad.

Reply:
I think the situation you're describing can't actually happen, the aliases lookup is only for getting a name for a block input from the loop inputs. With the SSA in place, any write will give it a new output name that wouldn't be found in the aliases map. I'm still not entirely convinced about this myself though, are you able to come up with a small concrete example that demonstrates this?

Reply:
Are you saying that the frontend will never generate core in this order? But don't you agree that it can generate:

```
While(cond, carry_init)
  graph (carry_in) {
    new_cond = Op2(carry_in) // now first
    carry_out = Op(carry_in) // now second
    return (new_cond, carry_out)
  }
```

from 

```
while cond:
  cond = carry_init * 2
  carry_init = carry_init * 2
```

And don't you agree that it is a valid optimization to reorder these two instructions? Then the bad graph can arise. Isn't it a design goal for the pretty-printer to work on optimized graphs?

Reply:
Ah, I hadn't considered re-ordering, I see what you're saying. What do you think about introducing a temporary if this situation is detected? For example for the script:

```
def while_test(a):
  b = a
  while a < 10:
    b = a * 2
    a = a + 2
  b = a + b
  return b
```

The IR could potentially look like:
```
prim::Loop(cond, a, b)
  block(a_in, b_in) {
    a_out = aten::add(a_in, 2)
    b_out = aten::mul(a_in, 2)
    new_cond = aten::lt(a_out, 10)
    ...
    -> (new_cond, a_out, b_out)
  }
```

And pretty printed:
```
t1 = a 			# added by pretty printer
while cond:
  t1 = a 		# added by pretty printer
  a = aten::add(t1, 2, 1)
  b = aten::mul(t1, 2)
  cond = aten::lt(a, 10)
return b
```